### PR TITLE
[pt1][quant] Fix the dimension mismatch issues when running the BERT model

### PR DESCRIPTION
**Pull request by jianyuh**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23139 [pt1][quant] Unit Test for the Dynamic Quantized Linear operator
* #24231 [pt1][quant] Add the default_weight_observer for the dynamic quantization path
* #24299 [pt1][quant] Remove the activation observer for default_qconfig
* #24232 [pt1][quant] Make the default qconfig_dict
* **#23330 [pt1][quant] Fix the dimension mismatch issues when running the BERT model**
* #23212 [pt1][quant] Add the type matching rule for qconfig_dict

We found the following dimension mismatch issues when running the BERT model with the dynamic quantization:
```
Traceback (most recent call last):
  File "bert.py", line 75, in <module>
    outputs = model(tokens_tensor, token_type_ids=segments_tensors)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 539, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/pytorch_transformers/modeling_bert.py", line 709, in forward
    head_mask=head_mask)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 539, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/pytorch_transformers/modeling_bert.py", line 437, in forward
    layer_outputs = layer_module(hidden_states, attention_mask, head_mask[i])
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 539, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/pytorch_transformers/modeling_bert.py", line 415, in forward
    attention_outputs = self.attention(hidden_states, attention_mask, head_mask)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 539, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/pytorch_transformers/modeling_bert.py", line 372, in forward
    self_outputs = self.self(input_tensor, attention_mask, head_mask)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 539, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/pytorch_transformers/modeling_bert.py", line 303, in forward
    query_layer = self.transpose_for_scores(mixed_query_layer)
  File "/home/jianyuhuang/anaconda3/lib/python3.7/site-packages/pytorch_transformers/modeling_bert.py", line 296, in transpose_for_scores
    return x.permute(0, 2, 1, 3)
RuntimeError: number of dims don't match in permute
```

Before the quantization, the dimension of ```x``` in ```transpose_for_scores``` is ```[1, 14, 12, 64]```;
After the quantization, the dimension of ```x``` in ```transpose_for_scores``` is ```[14, 12, 64]```.

There is a dimension mismatch on the output of the ```torch.ops.quantized.fbgemm_linear_dynamic``` operators. The first dimension is missing, which cause the issues with the abvove permute.

Differential Revision: [D16463334](https://our.internmc.facebook.com/intern/diff/D16463334/)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 1]
```diff
@@ -140,7 +140,11 @@ class QLinearInt8 final : public torch::OperatorKernel {
         /*thread_id=*/0,
         /*num_threads=*/1);
 
-    return output;
+    // The resulting matrix here is 2-D, let's view it with the original
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Why is this needed? qLinear works for Lenet without this change: can you provide more context?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please add tests to check for 3-D inputs.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 1 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: at::linear can take inputs of dimensions >2

it also highlights point about testing - we should see whether we can auto-convert relevant test_nn tests into quantized tests to catch situations like this.

In the meantime - can we please just add a unittest in this PR?

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 2]
```diff
@@ -140,7 +140,11 @@ class QLinearInt8 final : public torch::OperatorKernel {
...
+    // left hand dimensions of the input.
+    std::vector<int64_t> out_sizes = input.sizes().vec();
+    out_sizes.back() = N;
+    return output.view(out_sizes);
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp]: why do you need .view()? just allocate the tensor of the right shape above in _empty_affine_quantized call

**Review by dzhulgakov — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 1 above)
Comment (jianyuh) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: - Allocate the tensor of the right shape in the `at::zeros` function, instead of using `view` function.
- Add the comments with two examples to give more context and better elaboration for the cases we want to handle.
- Add the unit test in this PR.

**Comment by jianyuh** [comment]
> Please add tests to check for 3-D inputs.

Added the unit test for 3-D inputs.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 2 above)
Comment (jianyuh) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Thanks for pointing it out. Updated the code to address this issue.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 3]
```diff
@@ -120,9 +120,18 @@ class QLinearInt8 final : public torch::OperatorKernel {
         /*bias=*/bias_ptr,
         /*nCol=*/N);
 
+    // The resulting matrix here is 2-D, let's view it with the original
+    // left hand dimensions of the input. Here are two examples:
+    // 1. If the input tensor is {M, K}, the output tensor is {M, N}.
+    // 2. If the input tensor is {b, M, K}, the output tensor is {b, M, N}.
+    std::vector<int64_t> out_sizes = input.sizes().vec();
+    TORCH_CHECK(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp]: nit: this is redundant check - you already have a check on input rank above

File: aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp [hunk 4]
```diff
@@ -140,8 +140,17 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
         /*bias=*/bias_ptr,
         /*nCol=*/N);
 
+    // The resulting matrix here is 2-D, let's view it with the original
+    // left hand dimensions of the input. Here are two examples:
+    // 1. If the input tensor is {M, K}, the output tensor is {M, N}.
+    // 2. If the input tensor is {b, M, K}, the output tensor is {b, M, N}.
+    std::vector<int64_t> out_sizes = input.sizes().vec();
+    TORCH_CHECK(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp]: same

**Review by dzhulgakov — APPROVED** [review]

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 3 above)
Comment (jianyuh) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Will remove this.
